framework/JobHandler.py

```diff
@@ -542,26 +560,48 @@ def getFinished(self, removeFinished=True, jobIdentifier = '', uniqueHandler = "
       jobIdentifier = ''
 
     with self.__queueLock:
+      print('###############')
+      print(self.__batching)
+      print('@@ self.__finished --> ' + str(self.__finished))
       runsToBeRemoved = []
       for i,run in enumerate(self.__finished):
         ## If the jobIdentifier does not match or the uniqueHandler does not
         ## match, then don't bother trying to do anything with it
         if not run.identifier.startswith(jobIdentifier) \
-        or uniqueHandler != run.uniqueHandler:
+           or uniqueHandler != run.uniqueHandler:
           continue
+        ## check if the run belongs to a subgroup and in case
+        if run.groupId in self.__batching:
+          print('====> run.groupId in self.__batching: ' + str(run))
+          self.__batching[run.groupId]['finished'].append(run)
+        else:
+          finished.append(run)
+          print('====> run.groupId NOT in self.__batching: ' + str(run))
+          print('====> run.groupId' + str(run.groupId))
+          print(self.__batching)
 
-        finished.append(run)
         if removeFinished:
+          print('====> removeFinished: ' + str(removeFinished))
           runsToBeRemoved.append(i)
           self.__checkAndRemoveFinished(run)
+          ##FIXME: IF THE RUN IS PART OF A BATCH AND IT FAILS, WHAT DO WE DO? alfoa
+
+      ## check if batches are ready to be returned
+      for groupId in list(self.__batching.keys()):
+        if len(self.__batching[groupId]['finished']) ==  self.__batching[groupId]['size']:
+          doneBatch = self.__batching.pop(groupId)
+          finished.append(doneBatch['finished'])
```

Comment:
Hm, by moving finished.append to here, this might take quite a bit more time for it to be removed from various queues.  Possibly this might cause problems.

Reply:
@joshua-cogliati-inl : how would you suggest to edit it?

Reply:
The other problem here is that if getFinished is ever called with removeFinished == False, and there is a run in the batch, it will get added to self.__batching[run.groupId]['finished'], then when getFinished is called with removeFinished == True, that same run will be added to self.__batching[run.groupId]['finished'] again, and then this will mess up the len(self.__batching[groupId]['finished']) ==  self.__batching[groupId]['size']: check since can occur before it is done or it can exceed 'size' and never return any.  So getFinished will return the batch early or never return the batch if getFinished is ever called with removeFinished == False when there is a batch run in the __finished.

Reply:
Another comment is that it is worth considering extending the finished array instead of appending. As in:
```
finished.extend(doneBatch['finished'])
```
Then users of this function would not have to handle things like [run1,run2,[run3,run4,run5]].

Reply:
> The other problem here is that if getFinished is ever called with removeFinished == False, and there is a run in the batch, it will get added to self.__batching[run.groupId]['finished'], then when getFinished is called with removeFinished == True, that same run will be added to self.__batching[run.groupId]['finished'] again, and then this will mess up the len(self.__batching[groupId]['finished']) == self.__batching[groupId]['size']: check since can occur before it is done or it can exceed 'size' and never return any. So getFinished will return the batch early or never return the batch if getFinished is ever called with removeFinished == False when there is a batch run in the __finished.

I think this explains the randomness we observed.

Reply:
This is has been addressed by Congjian PR

Reply:
The append vs extend still needs to be discussed.

Reply:
I am fixing it. It is not needed